app/main.py

```diff
@@ -1,34 +1,108 @@
+            for row_index, col_index in coordinates
+        ]
 
     def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+        for deck in self.decks:
+            if row == deck.row and column == deck.col:
+                return deck
 
     def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+        deck = self.get_deck(row, column)
+        if deck.is_alive:
+            deck.is_alive = False
+
+        if not any(deck.is_alive for deck in self.decks):
+            self.is_drowned = True
+            return "Sunk!"
+
+        return "Hit!"
 
 
 class Battleship:
     def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
+        self.field = {}
+        for start, end in ships:
+            ship = Ship(start, end)
+            for deck in ship.decks:
+                self.field[deck.row, deck.col] = ship
+        self._validate_ships_amount()
+        self._validate_positions()
 
     def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+        if location in self.field:
+            ship = self.field[location]
+            return ship.fire(*location)
+        return "Miss!"
+
+    def print_field(self):
+        field = [
+            ["~" for _ in range(10)]
+            for _ in range(10)
+        ]
+        for (row, col), ship in self.field.items():
+            if ship.is_drowned:
+                field[row][col] = "x"
+                continue
+            deck = ship.get_deck(row, col)
+            field[row][col] = "□" if deck.is_alive else "*"
+
+        column_header = "   ".join([str(num) for num in range(10)])
+        print(f"    {column_header}")
+        formatted_rows = [
+            (f"{num}".center(4, " ") + "   ".join(row))
+            for num, row in enumerate(field)
+        ]
+        print("\n".join(formatted_rows))
+
+    def _validate_positions(self):
+        for (row, col), ship in self.field.items():
+            for row_delta in range(-1, 2):
+                for col_delta in range(-1, 2):
+                    check_coords = (row + row_delta, col + col_delta)
+                    ship_coords = [
+                        (deck.row, deck.col)
+                        for deck in ship.decks
+                    ]
+                    if (check_coords in self.field
+                            and check_coords not in ship_coords):
+                        raise Exception("Ships shouldn't be located"
+                                        "in the neighboring cells")
```

Comment:
Better create specific error and don't use Exception

Reply:
If I were writing a full implementation of the game, I would create a new BattleError parent class with all kinds of exceptions. But here I decided enough basic exception and description to it.